src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+                        if (matchedTypes[parameterIndex] && paramsMatchedByArray) return false;
+                        matchedTypes[parameterIndex] = true;
+                        paramsMatchedByArray = false;
+                    }
+                    else return false;
+
+                    if (targetArgument.Equals(arguments[i])) targetParamIndex = parameterIndex;
+                }
+                else return false;
+            }
+
+            // mark all optional parameters as matched
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                if (parameters[i].IsOptional || parameters[i].IsParams)
+                {
+                    matchedTypes[i] = true;
+                }
+            }
+
+            return Array.TrueForAll(matchedTypes, (item => item));
+        }
+
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var targetNodes = diagnostics.SelectAsArray(
+                d => root.FindNode(d.Location.SourceSpan, getInnermostNodeForTie: true)
+                         .GetAncestorsOrThis<ExpressionSyntax>().FirstOrDefault());
+
+            await editor.ApplyExpressionLevelSemanticEditsAsync(
+                document, targetNodes,
+                (semanticModel, targetNode) => true,
+                (semanticModel, currentRoot, targetNode) =>
+                {
+                    // All diagnostics have the same error code
+                    if (TryGetTargetTypeInfo(semanticModel, diagnostics[0].Id, targetNode, cancellationToken, out var nodeType, out var potentialConversionTypes)
+                        && potentialConversionTypes.Length == 1)
+                    {
+                        return ApplyFix(currentRoot, targetNode, potentialConversionTypes[0]);
+                    }
+
+                    return currentRoot;
+                },
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument, equivalenceKey: title)
+            {
+            }
+        }
+
+        private sealed class InheritanceDistanceComparer : IComparer<ITypeSymbol>
+        {
+            private readonly ITypeSymbol _baseType;
+            private readonly SemanticModel _semanticModel;
+
+            private int GetInheritanceDistance(ITypeSymbol baseType, ITypeSymbol? derivedType)
+            {
+                if (derivedType == null) return int.MaxValue;
+                if (derivedType.Equals(baseType)) return 0;
```

Comment:
When can this happen.  Wouldn't that mean we weren't adding a cast (since the types ar ethe same)?

Reply:
Type inheritance is like a graph, this is more like a BFS approach. This means we finally reach the base type that is the ancestor of the derived type which we passed in at the beginning. 
The same type conversion has been filtered before sorting, so the distance cannot be 0 except there is an explicit operator defined by users.